File: modules/expo-location-tracker/ios/ExpoLocationTrackerModule.swift
import ExpoModulesCore
import CoreLocation
import UIKit

public class ExpoLocationTrackerModule: Module, CLLocationManagerDelegate {
  private var locationManager: CLLocationManager?
  private var isTracking = false
  private var isPaused = false
  private var startTime: Date?
  private var lastLocation: CLLocation?
  private var currentConfig: LocationConfig?
  
  // Location filtering and smoothing
  private var locationBuffer: [CLLocation] = []
  private let bufferSize = 5
  private var lastValidLocation: CLLocation?
  
  // Background task management
  private var backgroundTaskIdentifier: UIBackgroundTaskIdentifier = .invalid
  
  public func definition() -> ModuleDefinition {
    Name("ExpoLocationTracker")
    
    Events("onLocationUpdate", "onTrackingStatusChange", "onPermissionStatusChange")
    
    AsyncFunction("startLocationUpdates") { (config: LocationConfig) in
      try await self.startLocationUpdates(config: config)
    }
    
    AsyncFunction("stopLocationUpdates") {
      try await self.stopLocationUpdates()
    }
    
    AsyncFunction("pauseLocationUpdates") {
      try await self.pauseLocationUpdates()
    }
    
    AsyncFunction("resumeLocationUpdates") {
      try await self.resumeLocationUpdates()
    }
    
    AsyncFunction("getCurrentLocation") { () -> [String: Any] in
      return try await self.getCurrentLocation()
    }
    
    AsyncFunction("requestPermissions") { () -> [String: Any] in
      return try await self.requestPermissions()
    }
    
    AsyncFunction("getTrackingStatus") { () -> [String: Any] in
      return self.getTrackingStatus()
    }
    
    OnAppEntersBackground {
      self.handleAppEntersBackground()
    }
    
    OnAppEntersForeground {
      self.handleAppEntersForeground()
    }
  }
  
  // MARK: - Public Methods
  
  private func startLocationUpdates(config: LocationConfig) async throws {
    self.currentConfig = config
    
    await MainActor.run {
      self.setupLocationManager()
      self.configureLocationManager(config: config)
      
      self.isTracking = true
      self.isPaused = false
      self.startTime = Date()
      
      self.locationManager?.startUpdatingLocation()
      
      // Enable background location if needed
      if config.backgroundTracking {
        self.locationManager?.allowsBackgroundLocationUpdates = true
        self.locationManager?.pausesLocationUpdatesAutomatically = false
      }
      
      self.sendTrackingStatusUpdate()
    }
  }
  
  private func stopLocationUpdates() async throws {
    await MainActor.run {
      self.locationManager?.stopUpdatingLocation()
      self.locationManager?.stopMonitoringSignificantLocationChanges()
      self.locationManager?.allowsBackgroundLocationUpdates = false
      
      self.endBackgroundTask()
      
      self.isTracking = false
      self.isPaused = false
      self.startTime = nil
      self.lastLocation = nil
      self.locationBuffer.removeAll()
      
      self.sendTrackingStatusUpdate()
    }
  }
  
  private func pauseLocationUpdates() async throws {
    await MainActor.run {
      self.isPaused = true
      self.sendTrackingStatusUpdate()
    }
  }
  
  private func resumeLocationUpdates() async throws {
    await MainActor.run {
      self.isPaused = false
      self.sendTrackingStatusUpdate()
    }
  }
  
  private func getCurrentLocation() async throws -> [String: Any] {
    return await withCheckedContinuation { continuation in
      DispatchQueue.main.async {
        self.setupLocationManager()
        
        self.locationManager?.requestLocation()
        
        // Store continuation for one-time location request
        self.oneTimeLocationContinuation = continuation
      }
    }
  }
  
  private var oneTimeLocationContinuation: CheckedContinuation<[String: Any], Never>?
  
  private func requestPermissions() async throws -> [String: Any] {
    return await withCheckedContinuation { continuation in
      DispatchQueue.main.async {
        self.setupLocationManager()
        
        let currentStatus = self.locationManager?.authorizationStatus ?? .notDetermined
        
        if currentStatus == .notDetermined {
          self.permissionContinuation = continuation
          self.locationManager?.requestAlwaysAuthorization()
        } else {
          let permissionStatus = self.createPermissionStatus(from: currentStatus)
          continuation.resume(returning: permissionStatus)
        }
      }
    }
  }
  
  private var permissionContinuation: CheckedContinuation<[String: Any], Never>?
  
  private func getTrackingStatus() -> [String: Any] {
    var status: [String: Any] = [
      "isTracking": isTracking,
      "isPaused": isPaused
    ]
    
    if let startTime = startTime {
      status["startTime"] = startTime.timeIntervalSince1970 * 1000
    }
    
    if let lastLocation = lastLocation {
      status["lastLocation"] = createLocationDict(from: lastLocation)
    }
    
    return status
  }
  
  // MARK: - Private Methods
  
  private func setupLocationManager() {
    if locationManager == nil {
      locationManager = CLLocationManager()
      locationManager?.delegate = self
    }
  }
  
  private func configureLocationManager(config: LocationConfig) {
    guard let manager = locationManager else { return }
    
    // Set accuracy based on config
    switch config.accuracy {
    case "high":
      manager.desiredAccuracy = kCLLocationAccuracyBest
    case "balanced":
      manager.desiredAccuracy = kCLLocationAccuracyNearestTenMeters
    case "low":
      manager.desiredAccuracy = kCLLocationAccuracyHundredMeters
    default:
      manager.desiredAccuracy = kCLLocationAccuracyBest
    }
    
    // Set distance filter
    manager.distanceFilter = config.distanceFilter
    
    // Configure for background usage
    if config.backgroundTracking {
      manager.allowsBackgroundLocationUpdates = true
      manager.pausesLocationUpdatesAutomatically = false
    }
  }
  
  private func createLocationDict(from location: CLLocation) -> [String: Any] {
    return [
      "latitude": location.coordinate.latitude,
      "longitude": location.coordinate.longitude,
      "altitude": location.altitude,
      "accuracy": location.horizontalAccuracy,
      "speed": location.speed >= 0 ? location.speed : nil,
      "heading": location.course >= 0 ? location.course : nil,
      "timestamp": location.timestamp.timeIntervalSince1970 * 1000,
      "source": "gps"
    ]
  }
  
  private func createPermissionStatus(from status: CLAuthorizationStatus) -> [String: Any] {
    let granted = status == .authorizedAlways || status == .authorizedWhenInUse
    let canAskAgain = status == .notDetermined
    
    var statusString: String
    switch status {
    case .authorizedAlways, .authorizedWhenInUse:
      statusString = "granted"
    case .denied:
      statusString = "denied"
    case .restricted:
      statusString = "restricted"
    case .notDetermined:
      statusString = "undetermined"
    @unknown default:
      statusString = "undetermined"
    }
    
    return [
      "granted": granted,
      "canAskAgain": canAskAgain,
      "status": statusString
    ]
  }
  
  private func isLocationValid(_ location: CLLocation) -> Bool {
    // Filter out locations with poor accuracy
    if location.horizontalAccuracy > 100 || location.horizontalAccuracy < 0 {
      return false
    }
    
    // Filter out locations that are too old
    if abs(location.timestamp.timeIntervalSinceNow) > 30 {
      return false
    }
    
    // Filter out impossible speeds (> 50 m/s = 180 km/h)
    if let lastLocation = lastValidLocation {
      let distance = location.distance(from: lastLocation)
      let timeInterval = location.timestamp.timeIntervalSince(lastLocation.timestamp)
      
      if timeInterval > 0 {
        let speed = distance / timeInterval
        if speed > 50 {
          return false
        }
      }
    }
    
    return true
  }
  
  private func smoothLocation(_ location: CLLocation) -> CLLocation? {
    locationBuffer.append(location)
    
    if locationBuffer.count > bufferSize {
      locationBuffer.removeFirst()
    }
    
    // Need at least 3 points for smoothing
    if locationBuffer.count < 3 {
      return location
    }
    
    // Simple moving average for smoothing
    let avgLat = locationBuffer.map { $0.coordinate.latitude }.reduce(0, +) / Double(locationBuffer.count)
    let avgLon = locationBuffer.map { $0.coordinate.longitude }.reduce(0, +) / Double(locationBuffer.count)
    let avgAlt = locationBuffer.compactMap { $0.altitude }.reduce(0, +) / Double(locationBuffer.compactMap { $0.altitude }.count)
    
    let smoothedCoordinate = CLLocationCoordinate2D(latitude: avgLat, longitude: avgLon)
    
    return CLLocation(
      coordinate: smoothedCoordinate,
      altitude: avgAlt,
      horizontalAccuracy: location.horizontalAccuracy,
      verticalAccuracy: location.verticalAccuracy,
      course: location.course,
      speed: location.speed,
      timestamp: location.timestamp
    )
  }
  
  private func sendTrackingStatusUpdate() {
    sendEvent("onTrackingStatusChange", getTrackingStatus())
  }
  
  // MARK: - Background Handling
  
  private func handleAppEntersBackground() {
    guard isTracking else { return }
    
    // Start background task to continue location updates
    startBackgroundTask()
    
    // Switch to significant location changes for better battery life
    if let locationManager = locationManager, currentConfig?.backgroundTracking == true {
      locationManager.startMonitoringSignificantLocationChanges()
    }
  }
  
  private func handleAppEntersForeground() {
    guard isTracking else { return }
    
    // End background task
    endBackgroundTask()
    
    // Resume normal location updates
    if let locationManager = locationManager {
      locationManager.stopMonitoringSignificantLocationChanges()
      locationManager.startUpdatingLocation()
    }
  }
  
  private func startBackgroundTask() {
    guard backgroundTaskIdentifier == .invalid else { return }
    
    backgroundTaskIdentifier = UIApplication.shared.beginBackgroundTask(withName: "LocationTracking") {
      // Background task is about to expire
      self.endBackgroundTask()
    }
  }
  
  private func endBackgroundTask() {
    guard backgroundTaskIdentifier != .invalid else { return }
    
    UIApplication.shared.endBackgroundTask(backgroundTaskIdentifier)
    backgroundTaskIdentifier = .invalid
  }
  
  // MARK: - CLLocationManagerDelegate
  
  public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
    guard let location = locations.last else { return }
    
    // Handle one-time location request
    if let continuation = oneTimeLocationContinuation {
      oneTimeLocationContinuation = nil
      let locationDict = createLocationDict(from: location)
      continuation.resume(returning: locationDict)
      return
    }
    
    // Only process if tracking and not paused
    guard isTracking && !isPaused else { return }
    
    // Validate location
    guard isLocationValid(location) else { return }
    
    // Apply smoothing if enabled
    let processedLocation: CLLocation
    if currentConfig?.adaptiveThrottling == true {
      processedLocation = smoothLocation(location) ?? location
    } else {
      processedLocation = location
    }
    
    lastLocation = processedLocation
    lastValidLocation = processedLocation
    
    let locationDict = createLocationDict(from: processedLocation)
    sendEvent("onLocationUpdate", locationDict)
  }
  
  public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
    print("Location manager failed with error: \(error.localizedDescription)")
    
    // Handle one-time location request failure
    if let continuation = oneTimeLocationContinuation {
      oneTimeLocationContinuation = nil
      let errorDict: [String: Any] = [
        "latitude": 0,
        "longitude": 0,
        "accuracy": -1,
        "timestamp": Date().timeIntervalSince1970 * 1000,
        "source": "error"
      ]
      continuation.resume(returning: errorDict)
    }
  }
  
  public func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
    let permissionStatus = createPermissionStatus(from: status)
    sendEvent("onPermissionStatusChange", permissionStatus)
    
    // Handle permission request continuation
    if let continuation = permissionContinuation {
      permissionContinuation = nil
      continuation.resume(returning: permissionStatus)
    }
  }
}

// MARK: - Supporting Types

struct LocationConfig: Record {
  @Field var accuracy: String = "high"
  @Field var interval: Int = 1000
  @Field var distanceFilter: Double = 0
  @Field var adaptiveThrottling: Bool = true
  @Field var backgroundTracking: Bool = true
}

File: modules/expo-location-tracker/android/src/main/java/expo/modules/locationtracker/ExpoLocationTrackerModule.kt
package expo.modules.locationtracker

import android.Manifest
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.os.Build
import android.os.IBinder
import android.os.Looper
import android.os.PowerManager
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationCompat
import androidx.core.content.ContextCompat
import com.google.android.gms.location.*
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import kotlinx.coroutines.*
import java.util.*
import kotlin.collections.ArrayList

class ExpoLocationTrackerModule : Module() {
  private var fusedLocationClient: FusedLocationProviderClient? = null
  private var locationCallback: LocationCallback? = null
  private var isTracking = false
  private var isPaused = false
  private var startTime: Long? = null
  private var lastLocation: Location? = null
  private var currentConfig: LocationConfig? = null
  
  // Location filtering and smoothing
  private val locationBuffer = ArrayList<Location>()
  private val bufferSize = 5
  private var lastValidLocation: Location? = null
  
  // Foreground service
  private var foregroundServiceIntent: Intent? = null
  
  override fun definition() = ModuleDefinition {
    Name("ExpoLocationTracker")
    
    Events("onLocationUpdate", "onTrackingStatusChange", "onPermissionStatusChange")
    
    AsyncFunction("startLocationUpdates") { config: LocationConfig ->
      startLocationUpdates(config)
    }
    
    AsyncFunction("stopLocationUpdates") {
      stopLocationUpdates()
    }
    
    AsyncFunction("pauseLocationUpdates") {
      pauseLocationUpdates()
    }
    
    AsyncFunction("resumeLocationUpdates") {
      resumeLocationUpdates()
    }
    
    AsyncFunction("getCurrentLocation") {
      getCurrentLocation()
    }
    
    AsyncFunction("requestPermissions") {
      requestPermissions()
    }
    
    AsyncFunction("getTrackingStatus") {
      getTrackingStatus()
    }
  }
  
  // MARK: - Public Methods
  
  private suspend fun startLocationUpdates(config: LocationConfig) = withContext(Dispatchers.Main) {
    currentConfig = config
    
    if (!hasLocationPermissions()) {
      throw Exception("Location permissions not granted")
    }
    
    setupLocationClient()
    
    val locationRequest = createLocationRequest(config)
    
    locationCallback = object : LocationCallback() {
      override fun onLocationResult(locationResult: LocationResult) {
        super.onLocationResult(locationResult)
        handleLocationUpdate(locationResult.lastLocation)
      }
      
      override fun onLocationAvailability(locationAvailability: LocationAvailability) {
        super.onLocationAvailability(locationAvailability)
        if (!locationAvailability.isLocationAvailable) {
          // Handle GPS signal loss
          sendEvent("onLocationUpdate", mapOf(
            "latitude" to 0.0,
            "longitude" to 0.0,
            "accuracy" to -1.0,
            "timestamp" to System.currentTimeMillis(),
            "source" to "unavailable"
          ))
        }
      }
    }
    
    try {
      fusedLocationClient?.requestLocationUpdates(
        locationRequest,
        locationCallback!!,
        Looper.getMainLooper()
      )
      
      // Start foreground service for background tracking
      if (config.backgroundTracking) {
        startForegroundService()
      }
      
      isTracking = true
      isPaused = false
      startTime = System.currentTimeMillis()
      
      sendTrackingStatusUpdate()
      
    } catch (securityException: SecurityException) {
      throw Exception("Location permission denied: ${securityException.message}")
    }
  }
  
  private suspend fun stopLocationUpdates() = withContext(Dispatchers.Main) {
    locationCallback?.let { callback ->
      fusedLocationClient?.removeLocationUpdates(callback)
    }
    
    stopForegroundService()
    
    isTracking = false
    isPaused = false
    startTime = null
    lastLocation = null
    locationBuffer.clear()
    
    sendTrackingStatusUpdate()
  }
  
  private suspend fun pauseLocationUpdates() = withContext(Dispatchers.Main) {
    isPaused = true
    sendTrackingStatusUpdate()
  }
  
  private suspend fun resumeLocationUpdates() = withContext(Dispatchers.Main) {
    isPaused = false
    sendTrackingStatusUpdate()
  }
  
  private suspend fun getCurrentLocation(): Map<String, Any?> = withContext(Dispatchers.Main) {
    if (!hasLocationPermissions()) {
      throw Exception("Location permissions not granted")
    }
    
    setupLocationClient()
    
    return@withContext suspendCancellableCoroutine { continuation ->
      try {
        fusedLocationClient?.lastLocation?.addOnSuccessListener { location ->
          if (location != null) {
            val locationMap = createLocationMap(location)
            continuation.resume(locationMap) {}
          } else {
            // Request a fresh location
            val locationRequest = LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, 1000)
              .setMaxUpdates(1)
              .build()
            
            val callback = object : LocationCallback() {
              override fun onLocationResult(locationResult: LocationResult) {
                super.onLocationResult(locationResult)
                val freshLocation = locationResult.lastLocation
                if (freshLocation != null) {
                  val locationMap = createLocationMap(freshLocation)
                  continuation.resume(locationMap) {}
                } else {
                  continuation.resume(mapOf(
                    "latitude" to 0.0,
                    "longitude" to 0.0,
                    "accuracy" to -1.0,
                    "timestamp" to System.currentTimeMillis(),
                    "source" to "error"
                  )) {}
                }
                fusedLocationClient?.removeLocationUpdates(this)
              }
            }
            
            fusedLocationClient?.requestLocationUpdates(
              locationRequest,
              callback,
              Looper.getMainLooper()
            )
          }
        }?.addOnFailureListener { exception ->
          continuation.resume(mapOf(
            "latitude" to 0.0,
            "longitude" to 0.0,
            "accuracy" to -1.0,
            "timestamp" to System.currentTimeMillis(),
            "source" to "error"
          )) {}
        }
      } catch (securityException: SecurityException) {
        continuation.resume(mapOf(
          "latitude" to 0.0,
          "longitude" to 0.0,
          "accuracy" to -1.0,
          "timestamp" to System.currentTimeMillis(),
          "source" to "error"
        )) {}
      }
    }
  }
  
  private suspend fun requestPermissions(): Map<String, Any> = withContext(Dispatchers.Main) {
    val context = appContext.reactContext ?: throw Exception("React context not available")
    
    val fineLocationGranted = ContextCompat.checkSelfPermission(
      context,
      Manifest.permission.ACCESS_FINE_LOCATION
    ) == PackageManager.PERMISSION_GRANTED
    
    val coarseLocationGranted = ContextCompat.checkSelfPermission(
      context,
      Manifest.permission.ACCESS_COARSE_LOCATION
    ) == PackageManager.PERMISSION_GRANTED
    
    val backgroundLocationGranted = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
      ContextCompat.checkSelfPermission(
        context,
        Manifest.permission.ACCESS_BACKGROUND_LOCATION
      ) == PackageManager.PERMISSION_GRANTED
    } else {
      true // Not required for older versions
    }
    
    val granted = fineLocationGranted && coarseLocationGranted && backgroundLocationGranted
    
    return@withContext mapOf(
      "granted" to granted,
      "canAskAgain" to true, // Android allows asking again
      "status" to if (granted) "granted" else "denied"
    )
  }
  
  private fun getTrackingStatus(): Map<String, Any?> {
    val status = mutableMapOf<String, Any?>(
      "isTracking" to isTracking,
      "isPaused" to isPaused
    )
    
    startTime?.let { time ->
      status["startTime"] = time
    }
    
    lastLocation?.let { location ->
      status["lastLocation"] = createLocationMap(location)
    }
    
    return status
  }
  
  // MARK: - Private Methods
  
  private fun setupLocationClient() {
    if (fusedLocationClient == null) {
      val context = appContext.reactContext ?: throw Exception("React context not available")
      fusedLocationClient = LocationServices.getFusedLocationProviderClient(context)
    }
  }
  
  private fun createLocationRequest(config: LocationConfig): LocationRequest {
    val priority = when (config.accuracy) {
      "high" -> Priority.PRIORITY_HIGH_ACCURACY
      "balanced" -> Priority.PRIORITY_BALANCED_POWER_ACCURACY
      "low" -> Priority.PRIORITY_LOW_POWER
      else -> Priority.PRIORITY_HIGH_ACCURACY
    }
    
    return LocationRequest.Builder(priority, config.interval.toLong())
      .setMinUpdateDistanceMeters(config.distanceFilter.toFloat())
      .setWaitForAccurateLocation(true)
      .build()
  }
  
  private fun hasLocationPermissions(): Boolean {
    val context = appContext.reactContext ?: return false
    
    val fineLocationGranted = ContextCompat.checkSelfPermission(
      context,
      Manifest.permission.ACCESS_FINE_LOCATION
    ) == PackageManager.PERMISSION_GRANTED
    
    val coarseLocationGranted = ContextCompat.checkSelfPermission(
      context,
      Manifest.permission.ACCESS_COARSE_LOCATION
    ) == PackageManager.PERMISSION_GRANTED
    
    return fineLocationGranted && coarseLocationGranted
  }
  
  private fun createLocationMap(location: Location): Map<String, Any?> {
    return mapOf(
      "latitude" to location.latitude,
      "longitude" to location.longitude,
      "altitude" to if (location.hasAltitude()) location.altitude else null,
      "accuracy" to location.accuracy.toDouble(),
      "speed" to if (location.hasSpeed()) location.speed.toDouble() else null,
      "heading" to if (location.hasBearing()) location.bearing.toDouble() else null,
      "timestamp" to location.time,
      "source" to "gps"
    )
  }
  
  private fun isLocationValid(location: Location?): Boolean {
    if (location == null) return false
    
    // Filter out locations with poor accuracy
    if (location.accuracy > 100 || location.accuracy <= 0) {
      return false
    }
    
    // Filter out locations that are too old
    val locationAge = System.currentTimeMillis() - location.time
    if (locationAge > 30000) { // 30 seconds
      return false
    }
    
    // Filter out impossible speeds (> 50 m/s = 180 km/h)
    lastValidLocation?.let { lastLocation ->
      val distance = location.distanceTo(lastLocation)
      val timeInterval = (location.time - lastLocation.time) / 1000.0 // seconds
      
      if (timeInterval > 0) {
        val speed = distance / timeInterval
        if (speed > 50) {
          return false
        }
      }
    }
    
    return true
  }
  
  private fun smoothLocation(location: Location): Location? {
    locationBuffer.add(location)
    
    if (locationBuffer.size > bufferSize) {
      locationBuffer.removeAt(0)
    }
    
    // Need at least 3 points for smoothing
    if (locationBuffer.size < 3) {
      return location
    }
    
    // Simple moving average for smoothing
    val avgLat = locationBuffer.map { it.latitude }.average()
    val avgLon = locationBuffer.map { it.longitude }.average()
    val avgAlt = locationBuffer.filter { it.hasAltitude() }.map { it.altitude }.average()
    
    val smoothedLocation = Location(location.provider)
    smoothedLocation.latitude = avgLat
    smoothedLocation.longitude = avgLon
    if (!avgAlt.isNaN()) {
      smoothedLocation.altitude = avgAlt
    }
    smoothedLocation.accuracy = location.accuracy
    smoothedLocation.time = location.time
    if (location.hasSpeed()) {
      smoothedLocation.speed = location.speed
    }
    if (location.hasBearing()) {
      smoothedLocation.bearing = location.bearing
    }
    
    return smoothedLocation
  }
  
  private fun handleLocationUpdate(location: Location?) {
    // Only process if tracking and not paused
    if (!isTracking || isPaused) return
    
    // Validate location
    if (!isLocationValid(location)) return
    
    // Apply smoothing if enabled
    val processedLocation = if (currentConfig?.adaptiveThrottling == true) {
      smoothLocation(location!!) ?: location
    } else {
      location!!
    }
    
    lastLocation = processedLocation
    lastValidLocation = processedLocation
    
    val locationMap = createLocationMap(processedLocation)
    sendEvent("onLocationUpdate", locationMap)
  }
  
  private fun startForegroundService() {
    val context = appContext.reactContext ?: return
    
    // Create notification channel for Android O and above
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      val channel = NotificationChannel(
        NOTIFICATION_CHANNEL_ID,
        "GPS Tracking",
        NotificationManager.IMPORTANCE_LOW
      ).apply {
        description = "Ongoing GPS tracking for trail running"
        setShowBadge(false)
      }
      
      val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
      notificationManager.createNotificationChannel(channel)
    }
    
    foregroundServiceIntent = Intent(context, LocationTrackingService::class.java)
    
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      context.startForegroundService(foregroundServiceIntent)
    } else {
      context.startService(foregroundServiceIntent)
    }
  }
  
  private fun stopForegroundService() {
    val context = appContext.reactContext ?: return
    foregroundServiceIntent?.let { intent ->
      context.stopService(intent)
    }
    foregroundServiceIntent = null
  }
  
  private fun sendTrackingStatusUpdate() {
    sendEvent("onTrackingStatusChange", getTrackingStatus())
  }
  
  companion object {
    private const val NOTIFICATION_CHANNEL_ID = "location_tracking_channel"
    private const val NOTIFICATION_ID = 1001
  }
}

// MARK: - Supporting Types

data class LocationConfig(
  @Field val accuracy: String = "high",
  @Field val interval: Int = 1000,
  @Field val distanceFilter: Double = 0.0,
  @Field val adaptiveThrottling: Boolean = true,
  @Field val backgroundTracking: Boolean = true
) : Record

// MARK: - Foreground Service

class LocationTrackingService : Service() {
  override fun onBind(intent: Intent?): IBinder? = null
  
  override fun onCreate() {
    super.onCreate()
    createNotificationChannel()
  }
  
  override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
    startForeground(NOTIFICATION_ID, createNotification())
    return START_STICKY
  }
  
  private fun createNotificationChannel() {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      val channel = NotificationChannel(
        NOTIFICATION_CHANNEL_ID,
        "GPS Tracking",
        NotificationManager.IMPORTANCE_LOW
      ).apply {
        description = "Ongoing GPS tracking for trail running"
        setShowBadge(false)
      }
      
      val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
      notificationManager.createNotificationChannel(channel)
    }
  }
  
  private fun createNotification(): Notification {
    val intent = packageManager.getLaunchIntentForPackage(packageName)
    val pendingIntent = PendingIntent.getActivity(
      this,
      0,
      intent,
      PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
    )
    
    return NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID)
      .setContentTitle("TrailRun GPS Tracking")
      .setContentText("Recording your trail run...")
      .setSmallIcon(android.R.drawable.ic_menu_mylocation)
      .setContentIntent(pendingIntent)
      .setOngoing(true)
      .setPriority(NotificationCompat.PRIORITY_LOW)
      .build()
  }
  
  companion object {
    private const val NOTIFICATION_CHANNEL_ID = "location_tracking_channel"
    private const val NOTIFICATION_ID = 1001
  }
}